h2o-algos/src/main/java/hex/tree/TreeHandler.java

```diff
@@ -0,0 +1,153 @@
+package hex.tree;
+
+import hex.genmodel.algos.tree.SharedTreeNode;
+import hex.genmodel.algos.tree.SharedTreeSubgraph;
+import hex.schemas.TreeV3;
+import water.MemoryManager;
+import water.api.Handler;
+
+import java.util.*;
+
+/**
+ * Handling requests for various model trees
+ */
+public class TreeHandler extends Handler {
+    private static final int NO_CHILD = -1;
+
+    public TreeV3 getTree(final int version, final TreeV3 args) {
+        final SharedTreeModel model = (SharedTreeModel) args.model.key().get();
+        if (model == null) throw new IllegalArgumentException("Given model does not exist: " + args.model.key().toString());
+        final SharedTreeModel.SharedTreeOutput sharedTreeOutput = (SharedTreeModel.SharedTreeOutput) model._output;
+        validateArgs(args, sharedTreeOutput);
+
+        final CompressedTree auxCompressedTree = sharedTreeOutput._treeKeysAux[args.tree_number][args.tree_class].get();
+        final SharedTreeSubgraph sharedTreeSubgraph = sharedTreeOutput._treeKeys[args.tree_number][args.tree_class].get()
+                .toSharedTreeSubgraph(auxCompressedTree, sharedTreeOutput._names, sharedTreeOutput._domains);
+
+        final TreeProperties treeProperties = convertSharedTreeSubgraph(sharedTreeSubgraph);
+
+        args.left_children = treeProperties._leftChildren;
+        args.right_children = treeProperties._rightChildren;
+        args.descriptions = treeProperties._descriptions;
+        args.root_node_id = sharedTreeSubgraph.rootNode.getNodeNumber();
+
+        return args;
+    }
+
+
+    /**
+     * @param args An instance of {@link TreeV3} input arguments to validate
+     * @param output An instance of {@link hex.tree.SharedTreeModel.SharedTreeOutput} to validate input arguments against
+     */
+    private void validateArgs(TreeV3 args, SharedTreeModel.SharedTreeOutput output) {
+        if (args.tree_number < 0) throw new IllegalArgumentException("Tree number must be greater than 0.");
+        if (args.tree_class < 0) throw new IllegalArgumentException("Tree class must be greater than 0.");
+
+        if (args.tree_number > output._treeKeys.length - 1)
+            throw new IllegalArgumentException("There is no such tree.");
+        if (args.tree_class > output._treeKeys[args.tree_number].length - 1)
```

Comment:
Binomial models have only a single tree per iteration.

Reply:
In case of Binomial and regression, name of the column is just ignored and that's it. The user always gets the first the, no matter what.